src/main/java/org/generationcp/middleware/api/germplasmlist/GermplasmListServiceImpl.java

```diff
@@ -74,4 +79,84 @@ public GermplasmListGeneratorDTO create(final GermplasmListGeneratorDTO request,
+	}
+
+	@Override
+	public Optional<GermplasmList> getGermplasmListByParentAndName(final String germplasmListName, final Integer parentId,
+		final String programUUID) {
+		return Optional.ofNullable(this.daoFactory.getGermplasmListDAO().getGermplasmListByParentAndName(germplasmListName, parentId, programUUID));
+	}
+
+	@Override
+	public Integer createGermplasmListFolder(final Integer userId, final String folderName, final Integer parentId,
+		final String programUUID) {
+
+		final GermplasmList parentFolder = (Objects.isNull(parentId)) ? null :
+			this.getGermplasmListById(parentId).orElseThrow(() -> new MiddlewareException("Parent Folder does not exist"));
+
+		final GermplasmList folder = new GermplasmList();
+		folder.setDate(Util.getCurrentDateAsLongValue());
+		folder.setUserId(userId);
+		folder.setDescription(folderName);
+		folder.setName(folderName);
+		folder.setNotes(null);
+		folder.setParent(parentFolder);
+		folder.setType(GermplasmList.FOLDER_TYPE);
+		folder.setProgramUUID(programUUID);
+		folder.setStatus(GermplasmList.Status.FOLDER.getCode());
+		return this.daoFactory.getGermplasmListDAO().save(folder).getId();
+	}
+
+	@Override
+	public Integer updateGermplasmListFolder(final Integer userId, final String folderName, final Integer folderId,
+		final String programUUID) {
+
+		final GermplasmList folder =
+			this.getGermplasmListById(folderId).orElseThrow(() -> new MiddlewareException("Folder does not exist"));
+
+		folder.setName(folderName);
+		folder.setDescription(folderName);
+
+		return this.daoFactory.getGermplasmListDAO().save(folder).getId();
+	}
+
+	@Override
+	public Integer moveGermplasmListFolder(final Integer germplasmListId, final Integer newParentFolderId, final boolean isCropList,
+		final String programUUID) {
+
+		final GermplasmList listToMove = this.getGermplasmListById(germplasmListId)
+			.orElseThrow(() -> new MiddlewareException("GermplasmList does not exist"));
+
+		final GermplasmList newParentFolder = (Objects.isNull(newParentFolderId)) ? null :
+			this.getGermplasmListById(newParentFolderId).orElseThrow(() -> new MiddlewareException("Parent Folder does not exist"));
+
+		// if the list is moved to the crop list, set the program uuid to null so that
+		// it will be accessible to all programs of the same crop.
+		if (isCropList) {
+			listToMove.setProgramUUID(null);
+		} else {
+			// else, just set the current programUUID
+			listToMove.setProgramUUID(programUUID);
+		}
+
+		listToMove.setParent(newParentFolder);
+		try {
+			return this.daoFactory.getGermplasmListDAO().saveOrUpdate(listToMove).getId();
+		} catch (final HibernateException e) {
+			throw new MiddlewareQueryException("Error in moveGermplasmList in GermplasmListServiceImpl: " + e.getMessage(), e);
+		}
+	}
+
+	@Override
+	public void deleteGermplasmListFolder(final Integer folderId) {
+		final GermplasmList folder = this.getGermplasmListById(folderId)
+			.orElseThrow(() -> new MiddlewareException("FolderId does not exist"));
```

Comment:
Please use MiddlewareRequestException and add the message in mw_messages_es.properties